drivers/platform/xilinx/timer.c

```diff
@@ -0,0 +1,385 @@
+#endif
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Initialize hardware timer and the handler structure associated with
+ *        it.
+ * @param [out] desc - Pointer to the reference of the device handler.
+ * @param [in] init_param - Initialization structure.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise
+ */
+int32_t timer_init(struct timer_desc **desc,
+		   struct timer_init_param *init_param)
+{
+	int32_t ret;
+	struct timer_desc *dev;
+	struct xil_timer_desc *xdesc;
+	struct xil_timer_init_param *xinit = init_param->extra;
+
+	if (!desc)
+		return FAILURE;
+	if (!init_param)
+		return FAILURE;
+
+	dev = (struct timer_desc *)calloc(1, sizeof(*dev));
+	if(!dev)
+		return FAILURE;
+	xdesc = (struct xil_timer_desc *)calloc(1, sizeof(*xdesc));
+	if(!xdesc)
+		goto error_desc;
+
+	dev->id = init_param->id;
+	dev->freq_hz = init_param->freq_hz;
+	dev->load_value = init_param->load_value;
+	dev->extra = xdesc;
+	xdesc->active_tmr = xinit->active_tmr;
+	xdesc->type = xinit->type;
+
+	switch (xdesc->type) {
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		xdesc->config = XScuTimer_LookupConfig(dev->id);
+		xdesc->instance = calloc(1, sizeof(XScuTimer));
+		ret = XScuTimer_CfgInitialize(xdesc->instance, xdesc->config,
+					      ((XScuTimer_Config *)xdesc->config)->BaseAddr);
+		if (ret != XST_SUCCESS)
+			goto error_xdesc;
+		XScuTimer_LoadTimer((XScuTimer *)xdesc->instance, dev->load_value);
+		XScuTimer_EnableAutoReload((XScuTimer *)xdesc->instance);
+		XScuTimer_EnableInterrupt((XScuTimer *)xdesc->instance);
+		break;
+#endif
+		goto error_xdesc;
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		ret = XTmrCtr_Initialize(xdesc->instance, dev->id);
+		if(ret != 0)
+			goto error_desc;
+		ret = XTmrCtr_SelfTest(xdesc->instance, dev->id);
+		if(ret != 0)
+			goto error_desc;
+		tmr_options = XTC_DOWN_COUNT_OPTION | XTC_INT_MODE_OPTION;
+		tmr_options |= dev->auto_reload ? XTC_AUTO_RELOAD_OPTION : 0;
+		XTmrCtr_SetOptions(xdesc->instance, xdesc->active_tmr,
+				   XTC_DOWN_COUNT_OPTION | XTC_INT_MODE_OPTION |
+				   XTC_AUTO_RELOAD_OPTION);
+		XTmrCtr_SetResetValue(xdesc->instance, xdesc->active_tmr,
+				      dev->load_value);
+		break;
+#endif
+		goto error_xdesc;
+	default:
+		goto error_xdesc;
+	}
+
+	*desc = dev;
+
+	return ret;
+error_xdesc:
+	free(xdesc);
+error_desc:
+	free(dev);
+
+	return ret;
+}
+
+/**
+ * @brief Free the memory allocated by timer_setup().
+ * @param [in] desc - Pointer to the device handler.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise
+ */
+int32_t timer_remove(struct timer_desc *desc)
+{
+	struct xil_timer_desc *xdesc = desc->extra;
+
+	if(!desc)
+		return FAILURE;
+
+	switch (xdesc->type) {
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		XScuTimer_Stop(xdesc->instance);
+		break;
```

Comment:
you should free(xdesc->instance) here

Reply:
I will remove it at the end with the other `free` instances.

Reply:
I suggested you to put it here because xdesc->instance allocation is only done in TIMER_PS

Reply:
Acknowledged. Will do.